tools/experiments/idle_ext_pmt.py

```diff
@@ -0,0 +1,255 @@
+        PMT_EDGECOUNTER: str, edgecounter device for PMT input.
+
+    This experiment assumes that the device has at least one AD9910 DDS and at least one TTL board.
+    """
+    DIFFERENTIAL_TRIGGER = None
+    REPUMP_AOM_CHANNELS = None
+    kernel_invariants = {
+        "DIFFERENTIAL_TRIGGER", "REPUMP_AOM_CHANNELS", "PMT_EDGECOUNTER", "repump_aoms",
+        "pmt_counter", "ad9910s", "ttl_outs"
+    }
+
+    def build(self):
+        super().build()
+        lowest_priority = -100  # we use priorities range from -100 to 100.
+        self.set_default_scheduling(priority=lowest_priority, pipeline_name="main")
+
+        self.rising_pulse = True
+        if self.DIFFERENTIAL_TRIGGER is None:
+            raise Exception("DIFFERENTIAL_TRIGGER must be defined.")
+        self.differential_trigger = self.get_device(self.DIFFERENTIAL_TRIGGER)
+        if self.REPUMP_AOM_CHANNELS is None:
+            raise Exception("REPUMP_AOM_CHANNELS must be defined.")
+        self.repump_aoms = [self.get_device(kk) for kk in self.REPUMP_AOM_CHANNELS]
+        self._get_all_dds_and_ttl_objects()
+
+    def prepare(self):
+        super().prepare()
+        self._get_repump_aom_states()
+
+    def host_startup(self):
+        pass
+
+    def host_cleanup(self):
+        self.disconnect_labrad()
+
+    @host_only
+    def _get_all_dds_and_ttl_objects(self):
+        self.ad9910s = []
+        for kk in self.devices.ad9910s:
+            self.ad9910s.append(self.get_device(kk))
+        self.ttl_outs = []
+        for kk in self.devices.ttl_outs:
+            self.ttl_outs.append(self.get_device(kk))
+
+    @host_only
+    def _get_repump_aom_states(self):
+        """Gets the states of repump AOMs.
+
+        If a repump AOM is set to off, don't turn it on during the differential mode sequence.
+        """
+        dds_params = _p.loads(self.cxn.artiq.get_dds_parameters())
+        self.repump_aom_states = []
+        for kk in self.REPUMP_AOM_CHANNELS:
+            self.repump_aom_states.append(dds_params[kk][-1])
+
+    @kernel
+    def kernel_before_loops(self):
+        self.core.reset()
+        self.core.break_realtime()
+
+    @kernel
+    def kernel_loop(self, loop_index: TInt64):
+        self.update_hardware()
+        self.core.break_realtime()
+
+        max_wait_time_mu = self.get_max_cycle()
+        total_trigger_wait_time_mu = 0
+        trigger_time = -1
+
+        # we will go through the whole time in 50 us intervals
+        trigger_cycle_mu = self.exp.core.seconds_to_mu(50 * us)
+
+        self.differential_trigger.count(now_mu())  # clears all existing timestamps.
+
+        if self.rising_pulse:
+            self._gate_func = self.differential_trigger.gate_rising_mu
+        else:
+            self._gate_func = self._trigger_ttl.gate_falling_mu
+
+        # waits for a trigger for trigger_cycle_mu.
+        self._gate_func(trigger_cycle_mu)
+        total_trigger_wait_time_mu += trigger_cycle_mu
```

Comment:
My feeling is these three lines can be removed.

Reply:
If we remove them, will that cause an error when we do trigger_time = self.differential_trigger.timestamp_mu() for the first time?

Reply:
Yes you are correct. We should probably set trigger_time to be 0 initially.